Máster Universitario en Inteligencia Artificial  
Universidad Internacional de Valencia  
03MIAR: Algoritmos de Optimización  
📌 Daniel Felipe Hernández Mancipe

---

# 🛠 Proyecto: Trabajo Práctico

URL: [GitHub](https://github.com/danielhndz/03MIAR-Algoritmos-de-Optimizacion/blob/main/Project/Seminario_Algoritmos.ipynb)

### Problemas:

> 1. [ ] Sesiones de doblaje
> 2. [ ] Organizar los horarios de partidos de La Liga
> 3. [x] **Combinar cifras y operaciones**

### Descripción del problema:

- El problema consiste en **analizar** el siguiente problema y **diseñar** un algoritmo que lo **resuelva**.
- Disponemos de las 9 cifras ($c$) del 1 al 9 (excluimos el cero) y de los 4 signos ($s$) básicos de las operaciones fundamentales: suma ($+$), resta ($-$), multiplicación ($\cdot$), y división ($/$).
- Debemos **combinarlos alternativamente sin repetir ninguno de ellos** para obtener una cantidad dada. Un ejemplo sería para obtener el 4:
  $$4 + 2 - \frac{6}{3} \cdot 1 = 4$$

1. Debe analizarse el problema para encontrar todos los **valores enteros** posibles planteando las siguientes cuestiones:
   1. ¿Qué valor **máximo y mínimo** se pueden obtener según las condiciones del problema?

      **Matiz importante sobre el formato de la expresión**: hay dos interpretaciones válidas del enunciado.
      - **Forma estricta**: la expresión debe tener exactamente el patrón $d_1\,s_1\,d_2\,s_2\,d_3\,s_3\,d_4\,s_4\,d_5$ (empieza con cifra y alterna 5 cifras y 4 signos).
      - **Forma extendida**: además se permiten expresiones que empiezan con signo $-$ unario, por ejemplo $-d_1\,s_2\,d_2\,s_3\,d_3\,s_4\,d_4$.

      Con este matiz, el valor **máximo** coincide en ambos casos y el **mínimo** cambia en una unidad:

      **Caso A (forma estricta):**

      $$
      9 \cdot 8 + 7 - \frac{6}{3} = 72 + 7 - 2 = \boxed{77}
      $$

      $$
      1 + \frac{4}{2} - 8 \cdot 9 = 1 + 2 - 72 = \boxed{-69}
      $$

      **Caso B (forma extendida con signo $-$ unario inicial):**

      $$
      9 \cdot 8 + \frac{7}{1} - 2 = 72 + 7 - 2 = \boxed{77}
      $$

      $$
      -9 \cdot 8 / 1 + 2 = -72 + 2 = \boxed{-70}
      $$

   2. ¿Es posible encontrar **todos los valores enteros posibles** entre dicho mínimo y máximo?

      Sí, en ambos casos (estricto y extendido) se obtienen todos los enteros del intervalo entre el mínimo y el máximo.

      **Caso A (forma estricta):**
      - Mínimo = $-69$, máximo = $77$.
      - Tamaño del intervalo entero esperado:
        $$
        |[-69,77] \cap \mathbb{Z}| = 77 - (-69) + 1 = 147
        $$
      - Valores enteros únicos obtenidos por fuerza bruta: $147$.
      - Conclusión: **sí** se obtienen todos los enteros del intervalo.

      **Caso B (forma extendida con signo $-$ unario inicial):**
      - Mínimo = $-70$, máximo = $77$.
      - Tamaño del intervalo entero esperado:
        $$
        |[-70,77] \cap \mathbb{Z}| = 77 - (-70) + 1 = 148
        $$
      - Valores enteros únicos obtenidos por fuerza bruta: $148$.
      - Conclusión: **sí** se obtienen todos los enteros del intervalo.

(\*) La respuesta es obligatoria

2. (\*) ¿Cuantas posibilidades hay sin tener en cuenta las restricciones?

Sin las restricciones, se podrían combinar las cifras ($c$) y los signos ($s$) con repetición, es decir, en cada paso de la construcción de la expresión, se tienen las mismas posibilidades de elección tanto para cifras como para signos. Por lo tanto, el número de posibilidades sin tener en cuenta las restricciones es:

$$
c \cdot s \cdot c \cdot s \cdot c \cdot s \cdot c \cdot s \cdot c =
$$

$$
c^5 \cdot s^4 = 9^5 \cdot 4^4 = \boxed{15,116,544}
$$

3. ¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones?

Con el mismo matiz de las preguntas 1.1 y 1.2, hay dos recuentos válidos según cómo se interprete la estructura de la expresión.

**Caso A (forma estricta: $d_1\,s_1\,d_2\,s_2\,d_3\,s_3\,d_4\,s_4\,d_5$):**

- Selección ordenada de 5 cifras distintas de 9: $P(9,5)=9\cdot8\cdot7\cdot6\cdot5$.
- Selección ordenada de los 4 signos sin repetición: $P(4,4)=4\cdot3\cdot2\cdot1$.

$$
N_A = P(9,5)\cdot P(4,4)
$$

$$
N_A = (9\cdot8\cdot7\cdot6\cdot5)\,(4\cdot3\cdot2\cdot1)=9!=\boxed{362,880}
$$

**Caso B (forma extendida, añadiendo expresiones con signo $-$ unario al inicio):**

Además del caso estricto, se agregan expresiones del tipo $-d_1\,s_2\,d_2\,s_3\,d_3\,s_4\,d_4$.

- Selección ordenada de 4 cifras distintas de 9: $P(9,4)=9\cdot8\cdot7\cdot6$.
- Permutación de los 3 signos restantes ($+,*,/$): $P(3,3)=3!=6$.

$$
N_{extra}=P(9,4)\cdot P(3,3)=9\cdot8\cdot7\cdot6\cdot3!=\boxed{18,144}
$$

$$
N_B = N_A + N_{extra}=362,880+18,144=\boxed{381,024}
$$

Por lo tanto:

- Si se usa la **forma estricta**, hay **362,880** posibilidades.
- Si se usa la **forma extendida**, hay **381,024** posibilidades.

Como referencia empírica del algoritmo de fuerza bruta:

- Caso A: $90,000$ expresiones con resultado entero.
- Caso B: $94,824$ expresiones con resultado entero.

### Modelo para el espacio de soluciones

4. (\*) ¿Cuál es la estructura de datos que mejor se adapta al problema? Arguméntalo. (Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguméntalo)

Las estructuras de datos que mejor se adaptan al problema son los `sets`, las `tuples`, un `dictionary`, y un `string`.

- Los `sets` se usan para:
  - Almacenar los dígitos
  - Almacenar los signos de operación
  - Almacenar las permutaciones de los dígitos que ya se han procesado en el caso de expresiones empezando con el operador $-$, esto es debido a que estas expresiones no tienen 5 dígitos sino 4, por lo tanto al iterar todas las permutaciones de dígitos (que tienen 5 dígitos), se repiten muchas veces los primeros 4. De esta manera al almacenar los primeros 4 dígitos de las permutaciones que ya se han procesado, se evita la repetición antes mencionada.
  - Almacenar los resultados de evaluar las expresiones construidas, para posteriormente hacer la diferencia con el conjunto de todos los valores enteros entre el mínimo y máximo obtenidos, y de esta manera saber si el algoritmo pudo generarlos.
  - Almacenar todos los valores enteros entre el mínimo y el máximo obtenidos por el algoritmo, para el propósito descrito en el inciso anterior.
- Se utilizan `sets` en lugar de `lists` por ser computacionalmente más eficientes.
- Las `tuples` las usa la librería `itertools` para almacenar cada una de las permutaciones, luego estas permutaciones se almacenan en un objeto de tipo `itertools.permutations`.
- El `dictionary` se usa para almacenar las expresiones generadas como `key` y el resultado de evaluar dicha expresión como `value` de dicha `key`. La razón es la facilidad que ofrece esta estructura para manterner un enlace entre la expresión y el resultado de evaluarla, sumado a su eficiencia computacional.
- El `string` se utiliza para construir cada una de las expresiones, evaluarlas con la función `eval`, y luego almacenarlas en el `dictionary` del inciso anterior. La razón es que básicamente es la estructura más inmediata y simple para representar una expresión, sumado a que es uno de los tipos (el otro es `code object`) esperados por la función `eval`.
- Aunque no se menciona por que no se usa directamente para resolver el problema, se utiliza una `list` (posteriormente ordenada) para mostrar los valores enteros entre el mínimo y el máximo obtenidos que el algoritmo no pudo obtener. El algoritmo nunca ejecuta esta línea ya que si obtiene todos los valores enteros entre el mínimo y el máximo. La razón es la necesidad (aunque no es necesariamente una necesidad, sino más bien una preferencia) de mostrar los valores ordenados.

### Según el modelo para el espacio de soluciones

5. (\*) ¿Cuál es la función objetivo?

En sentido estricto, este trabajo no está planteado como una optimización clásica de una sola función escalar, sino como un problema de **búsqueda exhaustiva / enumeración** de expresiones factibles.

Sea $x$ una expresión candidata. Su evaluación se define como:

$$
E(x)=d_1\;s_1\;d_2\;s_2\;d_3\;s_3\;d_4\;s_4\;d_5
$$

(y, en la variante extendida, también se admiten expresiones con $-$ unario inicial).

El conjunto factible $\mathcal{X}$ queda determinado por:

- Dígitos sin repetición: $d_i \in \{1,\dots,9\}$ y $d_i \neq d_j$ para $i\neq j$.
- Operadores sin repetición: $s_k \in \{+, -, \cdot, /\}$ y $s_k \neq s_\ell$ para $k\neq \ell$.
- Evaluación válida respetando precedencia estándar.
- Filtro de salida: conservar solo resultados enteros, $E(x) \in \mathbb{Z}$.

Por tanto, la “función objetivo” operativa del trabajo es **construir y evaluar** todas las $x\in\mathcal{X}$ para obtener el conjunto de resultados enteros:

$$
\mathcal{R}=\{E(x)\in\mathbb{Z}: x\in\mathcal{X}\}
$$

6. (\*) ¿Es un problema de maximización o minimización?

No en su formulación principal. Es un problema de **factibilidad + enumeración** (o CSP discreto), no de maximización/minimización directa.

- Se acepta o rechaza cada expresión según cumpla restricciones.
- El algoritmo recorre el espacio de búsqueda y agrega los resultados válidos.

Si se quisiera forzar una formulación de optimización, podría reinterpretarse como una **maximización de cobertura** (maximizar $|\mathcal{R}|$), pero ese no es el objetivo original del enunciado ni del algoritmo implementado.

### Fuerza bruta

7. Diseña un algoritmo para resolver el problema por fuerza bruta.

Se implementa la función `brute_force_alg(lim=LIMIT, strict_form=False)` para recorrer exhaustivamente las permutaciones de cifras y operadores y evaluar todas las expresiones válidas.

- Con `strict_form=True` (**forma estricta**), solo se generan expresiones del tipo $d_1\,s_1\,d_2\,s_2\,d_3\,s_3\,d_4\,s_4\,d_5$.
- Con `strict_form=False` (**forma extendida**), además se incluyen expresiones con signo $-$ unario inicial.

En ambos casos, el algoritmo conserva únicamente expresiones con resultado entero y reporta: número de expresiones evaluadas, número de enteros obtenidos, mínimo, máximo y cobertura completa del intervalo $[\min,\max]$.

Para comparar ambos escenarios en la misma ejecución se realizan dos llamadas:

- `BRUTE_FORCE_RESULTS_EXTENDED = brute_force_alg(strict_form=False)`
- `BRUTE_FORCE_RESULTS_STRICT = brute_force_alg(strict_form=True)`


In [18]:
import time
import itertools
import textwrap


# Number of random targets used in helper experiments.
LIMIT = 10
# Digits allowed by the statement (1..9, no repetition inside an expression).
FIGURES = set(range(1, 10))
# Arithmetic operators allowed by the statement (no repetition inside an expression).
SIGNS = {"+", "-", "*", "/"}
# Cache for target runs: key=(mode, target), value=metadata dict.
BRUTE_FORCE_TARGET_CACHE = {}


def brute_force_alg(  # NOSONAR(S3776)
        lim: int = LIMIT,
        strict_form: bool = False,
        target=None,
        verbose: bool = True,
        return_metadata: bool = False):
    """Run brute force search for strict/extended form.

    Parameters
    ----------
    lim : int
        Maximum number of missing values shown in non-target runs.
    strict_form : bool
        True -> evaluate only strict expressions (d1 s1 d2 s2 d3 s3 d4 s4 d5).
        False -> include the extended unary-minus starting form.
    target : int | None
        If provided, stop on first integer expression equal to target.
    verbose : bool
        Whether to print the run summary.
    return_metadata : bool
        If True, return (results, metadata). Otherwise return only results.

    Returns
    -------
    dict | tuple[dict, dict]
        Dictionary expression -> integer result, and optional metadata.
    """
    counter = 0
    # Stores only expressions whose evaluated value is an integer.
    results = {}
    start = time.time()
    mode = "strict" if strict_form else "extended"

    target_found = False
    target_expression = ""
    evaluated_before_first_target = None

    for signs_p in itertools.permutations(SIGNS, 4):
        # In extended mode, avoid regenerating the same 4-digit unary-minus branch.
        starting_minus = set()
        for figs_p in itertools.permutations(FIGURES, 5):
            # Build strict expression: d1 s1 d2 s2 d3 s3 d4 s4 d5
            expression = ""
            for i in range(4):
                expression += str(figs_p[i]) + signs_p[i]
            expression += str(figs_p[-1])

            counter, found_target, found_expression = add_expression(
                expression, results, counter, target)
            if found_target:
                target_found = True
                target_expression = found_expression
                evaluated_before_first_target = counter
                break

            # Extended mode adds unary-minus expression when first sign is '-'.
            if not strict_form and signs_p[0] == "-":
                if figs_p[:-1] not in starting_minus:
                    starting_minus.add(figs_p[:-1])
                    expression = ""
                    for i in range(4):
                        expression += signs_p[i] + str(figs_p[i])

                    counter, found_target, found_expression = add_expression(
                        expression, results, counter, target)
                    if found_target:
                        target_found = True
                        target_expression = found_expression
                        evaluated_before_first_target = counter
                        break

        if target_found:
            break

    metadata = {
        "mode": mode,
        "target": target,
        "target_found": target_found,
        "target_expression": target_expression,
        "evaluated_before_first_target": evaluated_before_first_target,
        "evaluated_expressions": counter,
        "expressions_with_integer_result": len(results),
        "unique_integer_results": len(set(results.values()))
    }

    # Keep target-run stats so repeated target queries can reuse them.
    if target is not None:
        BRUTE_FORCE_TARGET_CACHE[(mode, target)] = metadata

    if verbose:
        print_results(start, counter, results, lim, mode, metadata)

    if return_metadata:
        return results, metadata
    return results


def add_expression(
        expression: str,
        results: dict,
        counter: int,
        target=None):
    """Evaluate one candidate expression and update counters/results.

    Notes
    -----
    - `counter` increments for every unique expression evaluated with eval.
    - `results` stores only integer-valued expressions.
    """
    # Skip duplicates created by mixed generation paths.
    if expression in results:
        return counter, False, ""

    # Safe here because expression is built from controlled digits/operators.
    result = eval(expression)
    counter += 1

    found_target = False
    if result.is_integer():
        integer_result = int(result)
        results[expression] = integer_result
        if target is not None and integer_result == target:
            found_target = True

    return counter, found_target, expression


def print_results(
        start: float,
        counter: int,
        results: dict,
        lim: int = LIMIT,
        mode: str = "extended",
        metadata=None):
    """Print a human-readable summary for brute-force runs."""
    finish = time.time()
    values_set = set(results.values())
    is_target_run = metadata is not None and metadata.get("target") is not None

    if is_target_run:
        print(textwrap.dedent(f"""
            Brute-force algorithm results ({mode} form):

            Run time = {finish - start:.3f} seconds
            Evaluated expressions until first target hit (inclusive) = {counter}
            Evaluated expressions with integer result = {len(results)}
            Evaluated expressions with unique integer result = {len(values_set)}"""))
    else:
        print(textwrap.dedent(f"""
            Brute-force algorithm results ({mode} form):

            Run time = {finish - start:.3f} seconds
            Evaluated expressions = {counter}
            Evaluated expressions with integer result = {len(results)}
            Evaluated expressions with unique integer result = {len(values_set)}"""))

    if values_set:
        min_value = min(values_set)
        max_value = max(values_set)
        # Retrieve one witness expression for min/max values.
        print(
            f"Minimum value = {[k for k, v in results.items() if v == min_value][0]} = {min_value}")
        print(
            f"Maximum value = {[k for k, v in results.items() if v == max_value][0]} = {max_value}")

        if not is_target_run:
            missing_values = set(range(min_value, max_value+1)) - values_set
            all_ints_between_min_max = "Yes" if len(
                missing_values) == 0 else "No"
            print(
                f"Did it get all the integers between the min and max values? {all_ints_between_min_max}")
            if len(missing_values) > 0:
                print(textwrap.dedent(f"""
                    First {lim} missing values for the full range:
                    {sorted(list(missing_values))[:lim]}
                """))

    if is_target_run:
        target = metadata["target"]
        if metadata["target_found"]:
            print(
                f"First expression found for target {target}: {metadata['target_expression']}")
        else:
            print(f"No expression found for target {target}.")


# Precompute full search spaces once for both modes.
BRUTE_FORCE_RESULTS_EXTENDED = brute_force_alg(strict_form=False)
BRUTE_FORCE_RESULTS_STRICT = brute_force_alg(strict_form=True)



Brute-force algorithm results (extended form):

Run time = 1.566 seconds
Evaluated expressions = 381024
Evaluated expressions with integer result = 94824
Evaluated expressions with unique integer result = 148
Minimum value = -8*9+2/1 = -70
Maximum value = 7+8*9-2/1 = 77
Did it get all the integers between the min and max values? Yes

Brute-force algorithm results (strict form):

Run time = 1.462 seconds
Evaluated expressions = 362880
Evaluated expressions with integer result = 90000
Evaluated expressions with unique integer result = 147
Minimum value = 1+4/2-8*9 = -69
Maximum value = 7+8*9-2/1 = 77
Did it get all the integers between the min and max values? Yes


### Complejidad

7. Calcula la complejidad del algoritmo por fuerza bruta.

El coste del algoritmo por fuerza bruta se determina por dos factores:

- Cuántas expresiones se generan.
- Cuánto cuesta construir/evaluar cada expresión.

Definamos:

- $d$: número de cifras disponibles,
- $s$: número de signos disponibles,
- $m$: número de operadores usados en cada expresión (en este problema, $m=4$ y por tanto se usan $m+1=5$ cifras).

**Forma estricta** ($d_1\,s_1\,d_2\,s_2\,d_3\,s_3\,d_4\,s_4\,d_5$):

$$
N_{strict}=P(s,m)\cdot P(d,m+1)
$$

Para el caso del proyecto ($d=9$, $s=4$, $m=4$):

$$
N_{strict}=P(4,4)\cdot P(9,5)=4!\cdot \frac{9!}{(9-5)!}=\boxed{362{,}880}
$$

**Forma extendida** (añadiendo expresiones con $-$ unario inicial):

$$
N_{extra}=P(d,m)\cdot P(s-1,m-1)
$$

En este problema:

$$
N_{extra}=P(9,4)\cdot P(3,3)=\boxed{18{,}144}
$$

Luego:

$$
N_{extended}=N_{strict}+N_{extra}=\boxed{381{,}024}
$$

Si el coste de construir y evaluar una expresión es $O(m)$ (o $O(1)$ cuando $m$ es constante), la complejidad temporal queda:

$$
T_{strict}=O\big(P(s,m)\cdot P(d,m+1)\cdot m\big)
$$

$$
T_{extended}=O\big((P(s,m)\cdot P(d,m+1)+P(s-1,m-1)\cdot P(d,m))\cdot m\big)
$$

Para esta instancia concreta ($d=9$, $s=4$, $m=4$), $m$ es constante, así que el tiempo queda dominado por el número de expresiones evaluadas:

- Forma estricta: $\Theta(362{,}880)$,
- Forma extendida: $\Theta(381{,}024)$.

El coste en memoria es $O(k)$, donde $k$ es el número de expresiones con resultado entero que se almacenan en el diccionario de resultados.

### Búsqueda de una o todas las expresiones que evalúen a un objetivo

Ya con el espacio de soluciones completo, simplemente se tendría que buscar en el diccionario que claves tienen como valor el resultado de interés. Por ejemplo, se pueden obtener todas las expresiones que evalúan en `3` y `4` y la primera que evalúa a `42` y comparar ambos casos (estricto y extendido):


In [19]:
def expressions_by_target(
        target,
        all_expressions=True,
        lim=LIMIT,
        mode="extended"):
    """Query expressions by target value from brute-force results.

    Parameters
    ----------
    target : int
        Integer target to search for.
    all_expressions : bool
        True -> list all expressions (or first `lim`).
        False -> show first expression and evaluation count until hit.
    lim : int | None
        Max number of expressions to print when all_expressions=True.
    mode : str
        "strict" or "extended".
    """
    mode = mode.lower()
    if mode == "strict":
        strict_form = True
        results = BRUTE_FORCE_RESULTS_STRICT
    elif mode == "extended":
        strict_form = False
        results = BRUTE_FORCE_RESULTS_EXTENDED
    else:
        raise ValueError("mode must be 'strict' or 'extended'")

    if all_expressions:
        expressions = [e for e, v in results.items() if v == target]

        if not expressions:
            print(f"No expressions found for target={target} in {mode} mode.")
            return expressions

        print()
        print(
            f"Total expressions that evaluate to {target} in {mode} mode: {len(expressions)}")
        if lim is None or len(expressions) <= lim:
            for expression in expressions:
                print(expression)
        else:
            print(f"Showing first {lim} expressions:")
            for expression in expressions[:lim]:
                print(expression)
        return expressions

    # Reuse target-run metadata if available; otherwise run a silent target search.
    cache_key = (mode, target)
    target_metadata = BRUTE_FORCE_TARGET_CACHE.get(cache_key)
    if target_metadata is None:
        _, target_metadata = brute_force_alg(
            strict_form=strict_form,
            target=target,
            verbose=False,
            return_metadata=True)

    print()
    if target_metadata["target_found"]:
        print(f"First expression that evaluates to {target} in {mode} mode:")
        print(target_metadata["target_expression"])
        print(
            "Expressions evaluated until first target hit (inclusive): "
            f"{target_metadata['evaluated_before_first_target']}")
        return [target_metadata["target_expression"]]

    print(f"No expression found for target={target} in {mode} mode.")
    return []


# Examples requested in the notebook statement.
EXPRESSIONS_3_STRICT = expressions_by_target(
    3, all_expressions=True, mode="strict")
EXPRESSIONS_4_EXTENDED = expressions_by_target(
    4, all_expressions=True, mode="extended")
EXPRESSION_42_EXTENDED = expressions_by_target(
    42, all_expressions=False, mode="extended")
EXPRESSION_42_STRICT = expressions_by_target(
    42, all_expressions=False, mode="strict")



Total expressions that evaluate to 3 in strict mode: 1952
Showing first 10 expressions:
2+1*3-8/4
2+1*4-9/3
2+3*1-8/4
2+4*1-9/3
3+1*2-8/4
3+1*4-8/2
3+2*1-8/4
3+2*4-8/1
3+4*1-8/2
3+4*2-8/1

Total expressions that evaluate to 4 in extended mode: 2208
Showing first 10 expressions:
2+1*4-6/3
2+1*5-9/3
2+4*1-6/3
2+5*1-9/3
3+1*4-6/2
3+1*5-8/2
3+2*4-7/1
3+2*5-9/1
3+4*1-6/2
3+4*2-7/1

First expression that evaluates to 42 in extended mode:
1+5*9-8/2
Expressions evaluated until first target hit (inclusive): 836

First expression that evaluates to 42 in strict mode:
1+5*9-8/2
Expressions evaluated until first target hit (inclusive): 836


En la salida de la celda anterior, para `target=42`, el conteo **"Expressions evaluated until first target hit (inclusive)"** puede aparecer igual en modo estricto y extendido. Esto ocurre cuando la primera expresión que alcanza 42 pertenece al conjunto estricto y se encuentra en la parte del recorrido que ambos modos comparten. El modo extendido solo añade ramas extra (expresiones que empiezan con `-` unario), pero esas ramas no llegan a explorarse si el objetivo ya fue encontrado antes.

### Diseño de un mejor algoritmo

8. (\*) Diseña un algoritmo que mejore la complejidad del algoritmo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta.

### Backtracking con poda fuerte exacta

La implementación combina **dos rutas** según el objetivo de la ejecución:

- **Ruta de objetivo (`target` definido):** backtracking con poda fuerte exacta.
- **Ruta de enumeración (`target=None`):** backtracking completo optimizado para reducir sobrecarga constante.

En ambos casos, la evaluación respeta precedencia aritmética usando el estado incremental `(total, term)` con `Fraction`:

- `+` y `-` cierran el término actual (`total <- total + term`) y crean un nuevo `term`.
- `*` y `/` actualizan solo `term`.
- En hoja, el valor final es `total + term`.

Se puede ver una representación de como se desarrollaría cada una de las posibilidades en este flujo en el siguiente diagrama:

                                       1
                          ┌────────────┼────────────┬────────────┐
                         [+]          [-]          [*]          [/]
                         / \           ...           ...           ...
                       2     6
                     /   \
                   [-]   [*] ...
                  /   \
                 3    ...
                  \
                   [*]
                 /  |  \
                4   5   6
               /    |    \
             [/]    ...    ...
            / | \
           5  6  7  ...

Para la ruta con `target`, la poda fuerte se aplica así:

1. Para cada estado parcial, se verifica si todavía es posible llegar al objetivo.
2. El sufijo restante se modela como transformación afín exacta:

   `valor_final = total + a * term + b`

3. Los pares `(a, b)` se precomputan con memoización en `suffix_transforms`, indexados por máscaras de bits de cifras y signos restantes.
4. Si no existe ningún par que satisfaga la igualdad con el `target`, la rama se poda inmediatamente.

Para la ruta `target=None`, no hay poda por objetivo, pero sí una optimización importante de implementación. La lógica es la misma (recorrer todo el árbol), pero con menor coste por nodo:

- **Bitmasks para estado restante**: en lugar de listas de disponibles, se usan enteros (`figures_mask`, `signs_mask`) donde cada bit indica si una cifra/signo sigue libre.
  Comprobación y actualización se hacen con operaciones bit a bit (`&`, `^`), que son más baratas que copiar listas y hacer `remove`.

- **DFS especializada para enumeración (`dfs_all`)**: se separa la ruta de enumeración de la ruta con objetivo (`dfs_target`).
  Esto evita evaluar, en cada nodo, condiciones de poda que no aplican cuando `target=None`.

- **Mutación in-place de tokens**: la expresión parcial se guarda en un buffer preasignado y se escribe por posición.
  Así se evita crear listas nuevas con `append/pop` continuos y se reduce la presión de memoria.

- **Conversión reutilizable de cifras**: se precalculan `DIGIT_STR` y `DIGIT_FRACTION` para no reconstruir `str(d)` ni `Fraction(d)` en cada expansión del árbol.

- **Menos asignaciones intermedias**: al trabajar con máscaras y variables escalares (`total`, `term`) se minimiza la creación de objetos temporales en cada llamada recursiva.

Estas mejoras **no cambian** el orden asintótico en `target=None` (hay que visitar todas las hojas), pero sí reducen el factor constante. Por eso se observa menor tiempo de ejecución práctico para la enumeración completa, manteniendo exactamente el mismo número de expresiones evaluadas.


In [20]:
import time
from fractions import Fraction
from functools import lru_cache


# Fixed operator order used by recursion and bitmask encoding.
OPS_ORDER = ('+', '-', '*', '/')
# Each operator is mapped to one bit so we can represent remaining operators as a mask.
OP_BITS = {op: 1 << i for i, op in enumerate(OPS_ORDER)}
# Fixed digit order used in both strict and extended modes.
DIGITS_ORDER = tuple(range(1, 10))
DIGIT_BITS = {digit: 1 << (digit - 1) for digit in DIGITS_ORDER}
# Precomputed string/fraction representations to avoid repeated conversions in recursion.
DIGIT_STR = {digit: str(digit) for digit in DIGITS_ORDER}
DIGIT_FRACTION = {digit: Fraction(digit) for digit in DIGITS_ORDER}


def figures_to_mask(figures):
    """Encode available digits as a bitmask."""
    mask = 0
    for fig in figures:
        # Turn on the bit corresponding to this digit in the mask.
        mask |= DIGIT_BITS[fig]
    return mask


def signs_to_mask(signs):
    """Encode available operators as a bitmask."""
    mask = 0
    for sign in signs:
        # Turn on the bit corresponding to this operator in the mask.
        mask |= OP_BITS[sign]
    return mask


@lru_cache(maxsize=None)
def suffix_transforms(figures_mask, signs_mask):
    """Return all affine transforms (a, b) such that final = total + a*term + b.

    The cache key is the pair of remaining masks. This allows exact feasibility
    checks for target search without recomputing suffix expansions repeatedly.
    """
    if signs_mask == 0:
        # No operators left: suffix contributes exactly the current term.
        return ((Fraction(1), Fraction(0)),)

    transforms = set()
    for op in OPS_ORDER:
        op_bit = OP_BITS[op]
        if not (signs_mask & op_bit):
            continue

        next_signs_mask = signs_mask ^ op_bit
        for fig in DIGITS_ORDER:
            fig_bit = DIGIT_BITS[fig]
            if not (figures_mask & fig_bit):
                continue

            next_figures_mask = figures_mask ^ fig_bit
            fig_fraction = DIGIT_FRACTION[fig]

            for a, b in suffix_transforms(next_figures_mask, next_signs_mask):
                # Compose transforms according to operator semantics.
                if op == '+':
                    transforms.add((Fraction(1), a * fig_fraction + b))
                elif op == '-':
                    transforms.add((Fraction(1), -a * fig_fraction + b))
                elif op == '*':
                    transforms.add((a * fig_fraction, b))
                else:  # op == '/'
                    transforms.add((a / fig_fraction, b))

    return tuple(transforms)


def can_reach_target(total, term, figures_mask, signs_mask, target_fraction):
    """Check whether any suffix transform can still hit target exactly."""
    for a, b in suffix_transforms(figures_mask, signs_mask):
        if total + a * term + b == target_fraction:
            return True
    return False


def apply_operation(total, term, sign, fig_fraction):
    """Apply operator with precedence-aware state update.

    The pair (total, term) encodes expression value as total + term.
    + / - close the current term; * / extend it.
    """
    if sign == '+':
        return total + term, fig_fraction
    if sign == '-':
        return total + term, -fig_fraction
    if sign == '*':
        return total, term * fig_fraction
    return total, term / fig_fraction


def solve_with_backtracking(
        figures,
        signs,
        target=None,
        strict_form=False):
    """Solve expression search with backtracking (strict or extended mode).

    Modes
    -----
    - strict: only d1 s1 d2 s2 d3 s3 d4 s4 d5
    - extended: strict + unary-minus starting branch

    Behavior
    --------
    - target is None: full enumeration with optimized DFS (`dfs_all`).
    - target provided: target search with exact feasibility pruning (`dfs_target`).
    """
    mode = "strict" if strict_form else "extended"
    results = {
        "counter": 0,
        "results": {},
        "target_found": False,
        "target_expression": ""
    }
    start = time.time()

    figures_mask = figures_to_mask(figures)
    signs_mask = signs_to_mask(signs)
    # Extended unary-minus branch can only use +,*,/ afterwards.
    extended_signs_mask = signs_mask & ~OP_BITS['-']
    target_fraction = Fraction(target) if target is not None else None

    def visit_leaf(tokens, used_length, total, term):
        """Evaluate one fully formed expression and update global counters."""
        final_value = total + term
        results["counter"] += 1
        if final_value.denominator == 1:
            integer_result = final_value.numerator
            expression_str = "".join(tokens[:used_length])
            results["results"][expression_str] = integer_result
            if target_fraction is not None and final_value == target_fraction:
                results["target_found"] = True
                results["target_expression"] = expression_str

    def dfs_all(tokens, position, total, term, rem_figures_mask, rem_signs_mask):
        """Enumerate all valid expressions (no target pruning)."""
        if rem_signs_mask == 0:
            visit_leaf(tokens, position, total, term)
            return

        for sign in OPS_ORDER:
            sign_bit = OP_BITS[sign]
            if not (rem_signs_mask & sign_bit):
                continue

            next_signs_mask = rem_signs_mask ^ sign_bit
            tokens[position] = sign

            for fig in DIGITS_ORDER:
                fig_bit = DIGIT_BITS[fig]
                if not (rem_figures_mask & fig_bit):
                    continue

                next_figures_mask = rem_figures_mask ^ fig_bit
                tokens[position + 1] = DIGIT_STR[fig]
                next_total, next_term = apply_operation(
                    total, term, sign, DIGIT_FRACTION[fig])
                dfs_all(
                    tokens,
                    position + 2,
                    next_total,
                    next_term,
                    next_figures_mask,
                    next_signs_mask
                )

    def dfs_target(tokens, position, total, term, rem_figures_mask, rem_signs_mask):
        """Search target with exact feasibility pruning on remaining suffix."""
        if results["target_found"]:
            return

        # Strong pruning: drop branch if no suffix can reach target.
        if not can_reach_target(
                total, term, rem_figures_mask, rem_signs_mask, target_fraction):
            return

        if rem_signs_mask == 0:
            visit_leaf(tokens, position, total, term)
            return

        for sign in OPS_ORDER:
            sign_bit = OP_BITS[sign]
            if not (rem_signs_mask & sign_bit):
                continue

            next_signs_mask = rem_signs_mask ^ sign_bit
            tokens[position] = sign

            for fig in DIGITS_ORDER:
                fig_bit = DIGIT_BITS[fig]
                if not (rem_figures_mask & fig_bit):
                    continue

                next_figures_mask = rem_figures_mask ^ fig_bit
                tokens[position + 1] = DIGIT_STR[fig]
                next_total, next_term = apply_operation(
                    total, term, sign, DIGIT_FRACTION[fig])
                dfs_target(
                    tokens,
                    position + 2,
                    next_total,
                    next_term,
                    next_figures_mask,
                    next_signs_mask
                )
                if results["target_found"]:
                    return

    def run_strict_root():
        """Start strict-mode trees from each possible first digit."""
        tokens = [''] * 9
        for first_fig in DIGITS_ORDER:
            fig_bit = DIGIT_BITS[first_fig]
            if not (figures_mask & fig_bit):
                continue

            rem_figures_mask = figures_mask ^ fig_bit
            tokens[0] = DIGIT_STR[first_fig]
            initial_term = DIGIT_FRACTION[first_fig]

            if target_fraction is None:
                dfs_all(tokens, 1, Fraction(0), initial_term,
                        rem_figures_mask, signs_mask)
            else:
                dfs_target(tokens, 1, Fraction(0), initial_term,
                           rem_figures_mask, signs_mask)
                if results["target_found"]:
                    return

    def run_extended_root():
        """Start extended unary-minus trees from each possible first digit."""
        tokens = [''] * 8
        tokens[0] = '-'
        for first_fig in DIGITS_ORDER:
            fig_bit = DIGIT_BITS[first_fig]
            if not (figures_mask & fig_bit):
                continue

            rem_figures_mask = figures_mask ^ fig_bit
            tokens[1] = DIGIT_STR[first_fig]
            initial_term = -DIGIT_FRACTION[first_fig]

            if target_fraction is None:
                dfs_all(tokens, 2, Fraction(0), initial_term,
                        rem_figures_mask, extended_signs_mask)
            else:
                dfs_target(tokens, 2, Fraction(0), initial_term,
                           rem_figures_mask, extended_signs_mask)
                if results["target_found"]:
                    return

    # Strict branch is always part of both modes.
    run_strict_root()

    # Extended adds unary-minus root only when needed.
    if not strict_form and (target_fraction is None or not results["target_found"]):
        run_extended_root()

    finish = time.time()
    integer_results = len(results["results"])
    unique_integer_values = len(set(results["results"].values()))
    print(f"""
    Backtracking algorithm results ({mode} form):

    Run time = {finish - start:.3f} seconds
    Total expressions evaluated: {results["counter"]}
    Total expressions with integer result: {integer_results}
    Total unique integer results found: {unique_integer_values}""")

    if target is not None:
        if results["target_found"]:
            print(
                f"{' '*4}Expression for target {target} found: {results['target_expression']}")
        else:
            print(f"No expression found for target {target}.")

    return results["results"]


# Shared inputs for examples in this notebook section.
figures = list(FIGURES)
signs = list(SIGNS)

print("Finding a specific target (e.g., 42) - extended form")
solve_with_backtracking(figures, signs, 42, strict_form=False)

print()
print("Finding a specific target (e.g., 42) - strict form")
solve_with_backtracking(figures, signs, 42, strict_form=True)

print()
print("Finding all possible results - extended form")
all_results_extended = solve_with_backtracking(figures, signs, strict_form=False)

print()
print("Finding all possible results - strict form")
all_results_strict = solve_with_backtracking(figures, signs, strict_form=True)


Finding a specific target (e.g., 42) - extended form

    Backtracking algorithm results (extended form):

    Run time = 0.178 seconds
    Total expressions evaluated: 1
    Total expressions with integer result: 1
    Total unique integer results found: 1
    Expression for target 42 found: 1+5*9-8/2

Finding a specific target (e.g., 42) - strict form

    Backtracking algorithm results (strict form):

    Run time = 0.011 seconds
    Total expressions evaluated: 1
    Total expressions with integer result: 1
    Total unique integer results found: 1
    Expression for target 42 found: 1+5*9-8/2

Finding all possible results - extended form

    Backtracking algorithm results (extended form):

    Run time = 0.938 seconds
    Total expressions evaluated: 381024
    Total expressions with integer result: 94824
    Total unique integer results found: 148

Finding all possible results - strict form

    Backtracking algorithm results (strict form):

    Run time = 0.893 seconds
    Tota

En la ejecución mostrada, el backtracking con poda fuerte mejora claramente la búsqueda de objetivo (`target=42`) frente a fuerza bruta: encuentra una expresión válida tras apenas **1** evaluación, mientras que fuerza bruta en modo target necesita evaluar **836** expresiones para lograr el primer acierto.

Cuando se piden **todos los resultados** (`target=None`), ambos algoritmos deben recorrer el mismo número de hojas por modo (estricto/extendido). La diferencia práctica pasa a estar en la implementación: la versión actual de backtracking usa una ruta optimizada con bitmasks y mutación in-place, lo que reduce sobrecarga aunque el conteo de expresiones evaluadas sea el mismo.

### Complejidad

9. (\*) Calcula la complejidad del algoritmo.

La complejidad depende del modo de uso.

Sea:

- $d$: número de cifras disponibles,
- $s$: número de signos disponibles,
- $m$: número de operadores por expresión (aquí, $m=4$).

Para **enumeración completa** (`target=None`), no hay poda por objetivo y se recorre todo el espacio:

- **Modo estricto**:
  $$
  T_{strict}=\Theta(P(s,m)\cdot P(d,m+1))
  $$
- **Modo extendido**:
  $$
  T_{extended}=\Theta(P(s,m)\cdot P(d,m+1)+P(s-1,m-1)\cdot P(d,m))
  $$

En esta instancia:

- estricto: $\Theta(362{,}880)$,
- extendido: $\Theta(381{,}024)$.

La optimización con bitmasks e in-place **no cambia** ese orden asintótico, pero sí reduce el factor constante (menos asignaciones/copias por nodo).

Para **búsqueda de objetivo** (`target` definido), se activa poda fuerte exacta (`can_reach_target` + `suffix_transforms`):

- en promedio, reduce mucho el árbol explorado,
- en peor caso, sigue acotado por el recorrido completo del modo.

Por tanto:

$$
T_{mode,target}=O(T_{mode})
$$

con mejora práctica significativa cuando la poda es efectiva.

Memoria:

- $O(k)$ para el diccionario de expresiones enteras almacenadas,
- más la caché de transformaciones del sufijo (por estados de máscaras de cifras/signos).

### Datos de entrada aleatorios

10. Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios.


In [21]:
import numpy as np


# Target sampling interval derived from strict-mode integer range.
MIN_VALUE = -69
MAX_VALUE = 77


def generate_random_input(size=LIMIT, low=MIN_VALUE, high=MAX_VALUE):
    """Generate random integer targets in [low, high]."""
    # randint upper bound is exclusive, so use high+1.
    return np.random.randint(low, high+1, size)


print(generate_random_input())


[-40  -7  43  21 -35  57  53 -47 -17  52]


### Aplicación del algoritmo

11. Aplica el algoritmo al juego de datos generado.


In [22]:
def compare_random_targets_with_bruteforce_and_backtracking(
        strict_form=False,
        size=LIMIT):
    """Compare brute force vs backtracking for random targets.

    For each sampled target, this function prints both algorithm summaries
    with the same mode (strict or extended) to make the comparison direct.
    """
    mode = "strict" if strict_form else "extended"
    targets = [int(value) for value in generate_random_input(size=size)]

    print(f"Random targets ({mode} form): {targets}")

    for i, target in enumerate(targets, start=1):
        print()
        print("=" * 72)
        print(f"Target {i}/{len(targets)} = {target} ({mode} form)")

        print()
        print("[Brute force]")
        brute_force_alg(strict_form=strict_form, target=target)

        print()
        print("[Backtracking]")
        solve_with_backtracking(figures, signs, target, strict_form=strict_form)


# Run random-target comparison in extended mode by default.
compare_random_targets_with_bruteforce_and_backtracking(strict_form=False)


Random targets (extended form): [0, -51, 67, 28, 67, 25, 48, 67, 67, 42]

Target 1/10 = 0 (extended form)

[Brute force]

Brute-force algorithm results (extended form):

Run time = 0.064 seconds
Evaluated expressions until first target hit (inclusive) = 15212
Evaluated expressions with integer result = 2952
Evaluated expressions with unique integer result = 85
Minimum value = 1+2*3/6-9 = -7
Maximum value = 7+8*9-2/1 = 77
First expression found for target 0: 1+2*6/3-5

[Backtracking]

    Backtracking algorithm results (extended form):

    Run time = 0.003 seconds
    Total expressions evaluated: 1
    Total expressions with integer result: 1
    Total unique integer results found: 1
    Expression for target 0 found: 1+2-4*6/8

Target 2/10 = -51 (extended form)

[Brute force]

Brute-force algorithm results (extended form):

Run time = 0.130 seconds
Evaluated expressions until first target hit (inclusive) = 32271
Evaluated expressions with integer result = 7985
Evaluated expressions wi

### Comparación en targets aleatorios

En la celda anterior se comparan, para cada `target` aleatorio generado, los resultados de **fuerza bruta** y **backtracking con poda fuerte** en el mismo modo (por defecto, modo extendido). La comparación muestra dos patrones consistentes:

- Ambos algoritmos encuentran expresiones válidas para los objetivos del muestreo cuando el objetivo pertenece al rango alcanzable.
- Para la **gran mayoría de casos**, el backtracking encuentra una expresión objetivo al **primer intento** (`Total expressions evaluated = 1`).

Este comportamiento ocurre principalmente por tres motivos:

1. La poda de factibilidad (`can_reach_target`) es exacta: solo se exploran ramas cuyo sufijo todavía puede llegar al objetivo.
2. El contador de expresiones se incrementa solo al evaluar hojas completas; muchas ramas se descartan antes de llegar a hoja y no cuentan como expresiones evaluadas.
3. El orden de exploración, combinado con la alta cobertura de objetivos alcanzables en el rango del problema, hace que con frecuencia la primera rama factible ya produzca el `target`.

La mejora no es una garantía universal para todos los objetivos ni para cualquier orden de recorrido, pero en las ejecuciones del notebook se observa una ventaja práctica clara frente a fuerza bruta en búsqueda de objetivo. Además, el tiempo de ejecución de ambos algoritmos está parcialmente optimizado por el uso de caché: en fuerza bruta mediante `BRUTE_FORCE_TARGET_CACHE` para reutilizar búsquedas por target, y en backtracking mediante `lru_cache` en `suffix_transforms` para reutilizar transformaciones de sufijos.

### Referencias

12. Enumera las referencias que has utilizado(si ha sido necesario) para llevar a cabo el trabajo.
    - Universidad Internacional de Valencia (VIU). (2025). Apuntes y recursos de la asignatura Algoritmos de Optimización. Material docente en el aula virtual.
    - Universidad Internacional de Valencia (VIU). (2025). Apuntes y recursos de la asignatura Python para la Inteligencia Artificial. Material docente en el aula virtual.
    - https://docs.python.org/3/library/itertools.html

### Estudio avanzado del problema

13. Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño.

Para avanzar en el estudio del problema, las siguientes líneas son especialmente relevantes:

- Escalado del problema y generalización.
  Analizar qué ocurre al aumentar el número de cifras disponibles, el número de operadores o permitir repetición controlada. Esto permite estudiar cómo cambian el espacio de búsqueda y el rendimiento de cada estrategia.

- Ordenación heurística de ramas.
  Priorizar operadores/cifras que históricamente acercan más rápido al objetivo puede reducir aún más el tiempo medio en búsquedas con `target`.

- Poda adicional para enumeración completa.
  Aunque en `target=None` no se puede evitar visitar todo el conjunto de soluciones, sí se puede seguir reduciendo coste constante con mejoras de implementación (estructuras más compactas, menor creación de objetos, recorrido más cache-friendly y paralelización por ramas raíz).

- Variantes del objetivo.
  Reformular el problema para optimizar criterios como: minimizar número de operaciones, minimizar magnitud de cifras usadas, o maximizar cobertura de un intervalo bajo restricciones adicionales.

- Evaluación experimental sistemática.
  Diseñar benchmarks reproducibles (mismo conjunto de targets, semillas fijas, varias corridas) para comparar robustamente fuerza bruta y backtracking en tiempo, expresiones evaluadas y estabilidad.
